clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -54,6 +57,48 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
         }
 
         this.skipNegotiate = skipNegotiate;
+
+        this.httpClient = new OkHttpClient.Builder()
+                .cookieJar(new CookieJar() {
+                    private List<Cookie> cookieList = new ArrayList<>();
+
+                    @Override
+                    public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
+                        for (Cookie cookie : cookies) {
+                            boolean replacedCookie = false;
+                            for (int i = 0; i < cookieList.size(); i++) {
+                                Cookie innerCookie = cookieList.get(i);
+                                if (cookie.name().equals(innerCookie.name()) && innerCookie.matches(url)) {
+                                    // We have a new cookie that matches an older one so we replace the older one.
+                                    cookieList.set(i, innerCookie);
+                                    replacedCookie = true;
+                                    break;
+                                }
+                            }
+                            if (!replacedCookie) {
+                                cookieList.add(cookie);
+                            }
+                        }
+                    }
+
+                    @Override
+                    public List<Cookie> loadForRequest(HttpUrl url) {
+                        List<Cookie> matchedCookies = new ArrayList<>();
+                        List<Cookie> expiredCookies = new ArrayList<>();
+                        for (Cookie cookie: cookieList) {
+                            if (cookie.expiresAt() < System.currentTimeMillis()) {
+                                expiredCookies.add(cookie);
+                            } else if (cookie.matches(url)) {
+                                matchedCookies.add(cookie);
+                            }
+                        }
+
+                        cookieList.removeAll(expiredCookies);
```

Comment:
Do you have to worry about any concurrency here? Will there ever be overlapping requests?

Reply:
Every implementation I've found looks like it assumes no overlaps, including one by one of the maintainers of the library https://gist.github.com/swankjesse/e283f2ddf264abd9a0d45bc80e7ec2d8

Reply:
For your signalr scenarios is that a valid assumption? Or is concurrency managed by the calling httpclient? Java List certainly isn't thread safe.

Reply:
We don't guarantee requests are called in sequence. We'll give this a look and see if we need a lock.

Reply:
From some simple testing it looks like we need a lock